src/flamegpu/gpu/CUDASimulation.cu

```diff
@@ -277,6 +278,149 @@ void CUDASimulation::exitFunctions() {
+                    }
+                }
+            }
+        }
+    }
+}
+
+
+void CUDASimulation::spatialSortAgent(const std::string& agentName, const std::string& state, const int mode) {
+    float radius;
+    detail::Dims<float> envMin;
+    detail::Dims<float> envMax;
+    detail::Dims<float> envWidth;
+    detail::Dims<unsigned int> gridDim;
+
+    try {
+        radius = host_api->environment.getProperty<float>("INTERACTION_RADIUS");
+        envMin = {host_api->environment.getProperty<float>("MIN_POSITION"), host_api->environment.getProperty<float>("MIN_POSITION"), host_api->environment.getProperty<float>("MIN_POSITION")};
+        envMax = {host_api->environment.getProperty<float>("MAX_POSITION"), host_api->environment.getProperty<float>("MAX_POSITION"), host_api->environment.getProperty<float>("MAX_POSITION")};
+        envWidth = {(envMax.x-envMin.x), (envMax.y-envMin.y), (envMax.z-envMin.z)};
+        gridDim = {static_cast<unsigned int>(ceilf(envWidth.x / radius)), static_cast<unsigned int>(ceilf(envWidth.y / radius)), static_cast<unsigned int>(ceilf(envWidth.z / radius))};
+    } catch (exception::InvalidEnvProperty&) {
+        std::cout << "WARNING: Please set the INTERACTION_RADIUS, MIN_POSITION and MAX_POSITION environment properties to enable spatial sorting\n";
+        this->setSortAgentsEveryNSteps(0);
+        return;
+    }
+
+    CUDAAgent& cuda_agent = getCUDAAgent(agentName);
+
+    // Any agent in this list is guaranteed to have x, y, z and fgpu2_reserved_bin_index vars - used in the computation of spatial hash
+    // TODO: User could supply alternatives to "x", "y", "z" to use alternative variables?
+    void* xPtr = cuda_agent.getStateVariablePtr(state, "x");
+    void* yPtr = cuda_agent.getStateVariablePtr(state, "y");
+    void* zPtr = mode == Agent3D ? cuda_agent.getStateVariablePtr(state, "z") : 0;
+    void* binIndexPtr = cuda_agent.getStateVariablePtr(state, "fgpu2_reserved_bin_index");
+
+    // Compute occupancy
+    int blockSize = 0;  // The launch configurator returned block size
+    int minGridSize = 0;  // The minimum grid size needed to achieve the // maximum occupancy for a full device // launch
+    int gridSize = 0;  // The actual grid size needed, based on input size
+    const unsigned int state_list_size = cuda_agent.getStateSize(state);
+    cudaOccupancyMaxPotentialBlockSize(&minGridSize, &blockSize, calculateSpatialHash, 0, state_list_size);
+
+    //! Round up according to CUDAAgent state list size
+    gridSize = (state_list_size + blockSize - 1) / blockSize;
+
+    unsigned int sm_size = 0;
+    unsigned int streamIdx = 0;
+#if !defined(SEATBELTS) || SEATBELTS
+    auto *error_buffer = this->singletons->exception.getDevicePtr(streamIdx, this->getStream(streamIdx));
+    sm_size = sizeof(error_buffer);
+#endif
+
+    // Launch kernel
+    calculateSpatialHash<<<gridSize, blockSize, sm_size, this->getStream(streamIdx) >>> (reinterpret_cast<float*>(xPtr),
```

Comment:
I thought sorting was going to use the pre-existing spatial hash from agents outputting a message?

Reply:
Would the pre-existing hash be available in the append case? (I'm just stabbing in the dark, i've not looked into this at all)

Reply:
Without looking at code, I think it should be in the spare scan/scatter buffers during message output. It may get sorted along with the messages, which would require agents to be sorted with their messages.